lib/template-loader.php

```diff
@@ -107,6 +107,42 @@ function gutenberg_find_template( $template_file ) {
 	if ( $template_query->have_posts() ) {
 		$template_posts            = $template_query->get_posts();
 		$_wp_current_template_post = array_shift( $template_posts );
+
+		// Build map of template slugs to their priority in the current hierarchy.
+		$slug_priorities = new stdClass();
+		foreach ( $slugs as $index => $slug ) {
+			$slug_priorities->{$slug} = $index;
+		}
+
+		// See if there is a theme block template with higher priority than the resolved template post.
+		$higher_priority_block_template_path     = null;
+		$higher_priority_block_template_priority = PHP_INT_MAX;
+		foreach ( glob( get_template_directory() . '/block-templates/*.html', 1 ) as $path ) {
+			$theme_block_template_priority = $slug_priorities->{basename( $path, '.html' )};
+			if (
+				$theme_block_template_priority < $higher_priority_block_template_priority &&
+				$theme_block_template_priority < $slug_priorities->{$_wp_current_template_post->post_name}
+			) {
+				$higher_priority_block_template_path     = $path;
+				$higher_priority_block_template_priority = $theme_block_template_priority;
+			}
+		}
+
+		// If there is, use it instead.
+		if ( isset( $higher_priority_block_template_path ) ) {
+			$post_name                 = basename( $path, '.html' );
+			$_wp_current_template_post = get_post(
+				wp_insert_post(
```

Comment:
I think that `gutenberg_find_template()` is perhaps not ideally implemented to be populating a `$_wp_current_template_post` variable with the underlying `WP_Post`. What if instead it just obtained the HTML content? We aren't using anything about the template post other than the content (and the slug). If we were to avoid having to rely on an actual `WP_Post` being created, then we could avoid having to actual `wp_insert_post()` in the first place, and a site could freely use template content data straight from the file system when there is no `wp_template` post in the database. I think this would work out much better.

Reply:
Consider this change:

```diff
diff --git a/lib/template-loader.php b/lib/template-loader.php
index 9a91c93d3..108ca95aa 100644
--- a/lib/template-loader.php
+++ b/lib/template-loader.php
@@ -80,7 +80,7 @@ function gutenberg_override_query_template( $template, $type, array $templates =
  * @return string Path to the canvas file to include.
  */
 function gutenberg_find_template( $template_file ) {
-	global $_wp_current_template_post, $_wp_current_template_hierarchy;
+	global $_wp_current_template_content, $_wp_current_template_hierarchy;
 
 	// Bail if no relevant template hierarchy was determined, or if the template file
 	// was overridden another way.
@@ -106,7 +106,7 @@ function gutenberg_find_template( $template_file ) {
 
 	if ( $template_query->have_posts() ) {
 		$template_posts            = $template_query->get_posts();
-		$_wp_current_template_post = array_shift( $template_posts );
+		$current_template_post = array_shift( $template_posts );
 
 		// Build map of template slugs to their priority in the current hierarchy.
 		$slug_priorities = new stdClass();
@@ -121,7 +121,7 @@ function gutenberg_find_template( $template_file ) {
 			$theme_block_template_priority = $slug_priorities->{basename( $path, '.html' )};
 			if (
 				$theme_block_template_priority < $higher_priority_block_template_priority &&
-				$theme_block_template_priority < $slug_priorities->{$_wp_current_template_post->post_name}
+				$theme_block_template_priority < $slug_priorities->{$current_template_post->post_name}
 			) {
 				$higher_priority_block_template_path     = $path;
 				$higher_priority_block_template_priority = $theme_block_template_priority;
@@ -130,18 +130,7 @@ function gutenberg_find_template( $template_file ) {
 
 		// If there is, use it instead.
 		if ( isset( $higher_priority_block_template_path ) ) {
-			$post_name                 = basename( $path, '.html' );
-			$_wp_current_template_post = get_post(
-				wp_insert_post(
-					array(
-						'post_content' => file_get_contents( $higher_priority_block_template_path ),
-						'post_title'   => ucfirst( $post_name ),
-						'post_status'  => 'auto_draft',
-						'post_type'    => 'wp_template',
-						'post_name'    => $post_name,
-					)
-				)
-			);
+			$_wp_current_template_content = file_get_contents( $higher_priority_block_template_path );
 		}
 	}
 
@@ -167,17 +156,15 @@ function gutenberg_render_title_tag() {
  * Renders the markup for the current template.
  */
 function gutenberg_render_the_template() {
-	global $_wp_current_template_post;
+	global $_wp_current_template_content;
 	global $wp_embed;
 
-	if ( ! $_wp_current_template_post || 'wp_template' !== $_wp_current_template_post->post_type ) {
+	if ( ! $_wp_current_template_content ) {
 		echo '<h1>' . esc_html__( 'No matching template found', 'gutenberg' ) . '</h1>';
 		return;
 	}
 
-	$content = $_wp_current_template_post->post_content;
-
-	$content = $wp_embed->run_shortcode( $content );
+	$content = $wp_embed->run_shortcode( $_wp_current_template_content );
 	$content = $wp_embed->autoembed( $content );
 	$content = do_blocks( $content );
 	$content = wptexturize( $content );
```

Reply:
With this in place, then the next thing would be to provide a way to start a new `wp_template` post in the DB based on the data coming from the theme. Perhaps when creating a `wp_tempalte` post there could be some dropdown UI for selecting either theme-provided template content to override, or even another existing `wp_template` post to use as the basis (e.g. if wanting to fork a `singular` to be a more specific `page` template).

Reply:
Conclusion based on our discussion in-person at contributor day: we'll only create the auto-draft post when in the context of the admin when intending to edit. Otherwise, on the frontend, the static content should be used instead to avoid doing DB inserts for requests to view the site.

Reply:
Also, if doing this we won't need to change the `WP_Query` in `gutenberg_find_template()` would not need to be changed include the `auto-draft` status. It can remain `publish` posts only because the `auto-draft` indicates it has not been edited.

Reply:
SGTM like that.

Reply:
https://github.com/WordPress/gutenberg/pull/18247/commits/f11d41c9f3a6f9825eebf086c117019536f6d65a